Add upm support

rogerbarton opened the pull request:
Users can then nicely add the repo as a submodule in the Packages folder and so get updates.

- [ ] Add main author to `packages.json` @mcpiroman 
- [x] Update readme
- [x] Adapt "Only Executing Assembly" to reference "Assembly-CSharp.dll", because of the `.asmdef`
- [x] Should `.meta` files be included (for consistent guids)? Yes

If users do not want to use upm, e.g. if they are on an older version, it should still work.

(The `"revision": "6c6b..."` is currently the lastest commit on the master branch)

rogerbarton committed:
Add upm support

Allow users to add this to the Packages folder for it to be used as an embedded package
Users can then nicely add the repo as a submodule

mcpiroman commented:
Well, admittedly, I've never heard of upm, but it looks really cool. (I guess it works similarly to other package managers I know ;)). So yes, thanks for bringing that in. The only thing I'm worried about are the .asmdef files. Are they necessary? Because as you pointed out, they may interfere with assembly resolution (and for this little code there is here, this seems like little overkill?). I'm nevertheless not familiar with that solution, so I'll rely on your words.

rogerbarton committed:
Add meta files

rogerbarton commented:
Yes upm is based on npm (it's been around since 2018.1 I believe). From what I know the you need an asmdef per package and a separate one for any editor code, which we have here. It is overkill to have them but it doesn't work without them.

rogerbarton committed:
Potential fix for assemblyPaths using their manifest name

rogerbarton committed:
Rename asmdefs

rogerbarton committed:
Fallback to windows #if

rogerbarton committed:
Improve finding assemblies, always search in our assembly

Renamed Editor asmdef

rogerbarton committed:
assemblyPaths small fixes, editor gui adapted

EditorGUI for the target assemblies is now just a string array. Removed code to find _allKnownAssemblies.
assemblyPaths renamed to assemblyNames is now a list.
By default only Assembly-CSharp and the Editor variant is targeted. The assembly containing this tool is always included so we find the attributes.

rogerbarton committed:
Improve UI, dropdowns, add _allKnownAssemblies back

Simplified finding all assemblies, moved to RefreshAllKnownAssemblies()
Unity Assemblies are ignored by default.
Option to add all known assemblies.
Increased refresh time to 5 seconds from 1.

rogerbarton commented:
a5fd4553cbb908905af607886c99a39c318b4a20 also links to https://github.com/mcpiroman/UnityNativeTool/issues/17#issuecomment-605617438 and implements the whitelisting of assemblies, by their name (without extension). By default `Assembly-CSharp` and `Assembly-CSharp-Editor` are included. Our two asmdefs are always included, and so searched for the attributes. 

(042c4aa383876d43b3eb6a52a6881029d27fef22 and 06abb19cdb524cd1cc17ae1f627f28152490437c) I've now adapted the GUI for selecting the assemblies, it should be quite similar to before with a dropdown. I've slightly modified how `_allKnownAssemblies` is calculated, so Unity assemblies are ignored (theres a `IGNORED_ASSEMBLY_PREFIXES` string[] for exact control). The `Find` dropdown allows you to select from this list. 
![image](https://user-images.githubusercontent.com/42413282/78666907-59052580-78d8-11ea-94f2-1e3d02b08d23.png)

rogerbarton committed:
Add gui headers, info box

rogerbarton commented:
Added some headers and reordered. Any suggestions?
![image](https://user-images.githubusercontent.com/42413282/78672534-10059f00-78e1-11ea-9878-a86b7e7d062e.png)

rogerbarton commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -174,79 +179,66 @@ private void DrawUsedDlls(IList<NativeDllInfo> usedDlls)
         private void DrawOptions(DllManipulatorOptions options)
         {
+            options.onlyInEditor = EditorGUILayout.Toggle(ONLY_IN_EDITOR, options.onlyInEditor);
+            options.enableInEditMode = EditorGUILayout.Toggle(ENABLE_IN_EDIT_MODE, options.enableInEditMode);
```
This line requires #12 to be merged

rogerbarton committed:
Update README.md

rogerbarton commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -8,21 +8,24 @@
 #endif
 using System.IO;
 using System;
+using UnityEditorInternal;
```
Required for `AssemblyDefinitionAsset` to find all asmdefs in the project in `RefreshAllKnownAssemblies()`. This is not the best idea, I'll try and remove the dependency...

rogerbarton replied:
Removed in a2a212cc2ba011dd8cbe43b0ccc22fd85fa7f742

rogerbarton committed:
Partial revert "assemblyPaths small fixes, editor gui adapted", remove UnityEditorInternal dependency

Use original code to find all assemblies, use `Path.GetFileNameWithoutExtension` instead of `PathUtils.NormallizeUnityAssemblyPath`.

rogerbarton committed:
Merge branch 'master' into upm-support

mcpiroman commented:
> Added some headers and reordered. Any suggestions?

Yea, I very like that,, maybe because I have thought about sth like this either :). 

Generally, please let me know when I can review this. Btw Github allows you to declare a PR as preview, idk if they have already enabled marking existing ones as preview, which was requested a long time ago.

rogerbarton commented:
Ah I didn't know about the marking as preview, I'll do it in the future. This PR and #15 aren't ready for review yet. I'll tell you when they are...

rogerbarton committed:
Use string[] for Options.assemblyNames

rogerbarton committed:
Revert "Use string[] for Options.assemblyNames"

This reverts commit 8e1d0e370bdbeb533879ec602e39a0995ff73283.

rogerbarton committed:
Changed package.json

rogerbarton committed:
Bug fix, disable whole gui in play mode

rogerbarton committed:
Bug fix, hide errors for not finding default assemblies

Some project may not have any editor scripts so will always get an error as Assembly-CSharp-Editor does not exist.

rogerbarton commented:
Once #15 is merged, this PR is also ready for review.

rogerbarton committed:
Merge branch 'master' into upm-support

rogerbarton committed:
Merge fixes

mcpiroman commented on package.json:
```diff
@@ -0,0 +1,21 @@
+  ],
+  "category": "Unity",
+  "author": {
+    "name": "John Doe",
```
If I recall correctly, that's not my name :)

mcpiroman replied:
Nor an email

rogerbarton replied:
Where can I find them? they're not in your public profile.

mcpiroman replied:
I think we can live without that unless that's somehow mandatory

rogerbarton replied:
I can remove the `author` field if you want it to stay anonymous? (fyi John Doe is just a placeholder name)

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -644,34 +654,34 @@ private static void WriteNativeCrashLog(NativeFunction nativeFunction, object[]
             return PInvokes_Linux.dlopen(filepath, (int)Options.posixDlopenFlags);
 #elif UNITY_STANDALONE_OSX
             return PInvokes_Osx.dlopen(filepath, (int)Options.posixDlopenFlags);
+#else // UNITY_STANDALONE_WIN
```
Is the way it was somehow bad? I understand the addition of `|| UNITY_EDITOR_XXX` (which should also go there), but not the commented else. I mean, if for any reason the running platform happens not to be listed here, the code should not even compile.

rogerbarton replied:
I though it might be useful in some cases as there's [quite a lot of platforms](https://docs.unity3d.com/Manual/PlatformDependentCompilation.html). But then it probably makes most sense to leave it as `#elif UNITY_STANDALONE_WIN || UNITY_EDITOR_WIN` as this is what is directly supported. I will change it to that.

mcpiroman commented on scripts/Editor/MCpiroman.UnityNativeTool.Editor.asmdef:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "MCpiroman.UnityNativeTool.Editor",
```
don't know if it matters, but my nick is not spelled in the same casing (if it does I prefer lowercase)

rogerbarton replied:
I thought I saw it written somewhere like that, I'll change it to all lowercase.

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -17,6 +17,18 @@ public partial class DllManipulator
+            , "Assembly-CSharp-Editor"
+            #endif
+        };
+        public static readonly string[] INTERNAL_ASSEMBLY_NAMES = {"MCpiroman.UnityNativeTool"
```
don't know if it matters, but my nick is not spelled in the same casing (if it does I prefer lowercase)

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -48,20 +60,18 @@ internal static void Initialize(int unityMainThreadId, string assetsPath)
+            assemblyPathsTemp = assemblyPathsTemp.Concat(INTERNAL_ASSEMBLY_NAMES);
+            
+            var allAssemblies = AppDomain.CurrentDomain.GetAssemblies();
+            var assemblies = allAssemblies.Where(a => !a.IsDynamic && assemblyPathsTemp.Any(p => p == Path.ChangeExtension(a.ManifestModule.Name, null))).ToArray();
```
Is `ManifestModule.Name` guaranteed to be file name without path? If not, wouldn't it be better to use `Path.GetFileNameWithoutExtension`?

rogerbarton replied:
Having looked at the docs "This is usually, but not necessarily, the file name..." so I'll change it to use the path.

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -313,7 +303,32 @@ private void DrawOptions(DllManipulatorOptions options)
+            _allKnownAssemblies = playerCompiledAssemblies
+                .Concat(editorCompiledAssemblies)
+                .Concat(assemblyAssets)
+                .Where(a => !DllManipulator.IGNORED_ASSEMBLY_PREFIXES.Any(a.StartsWith))
```
To be exact, this wouldn't be `allKnownAssemblies` if we exclude some here. So maybe rename that to sth like `possibleTargetAssemblies`?

rogerbarton replied:
makes sense, changed to `possibleTargetAssemblies`

rogerbarton committed:
Review changes

rogerbarton committed:
Removed upm author field

mcpiroman commented:
Thanks

rogerbarton commented:
Thanks for reviewing!! I hope its not too much of a pain. I've got two small PRs left and then that should be it for a while :D
